// @brunoc/ngx-viacep/esm2015/lib/ngx-viacep.service.js
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { of } from 'rxjs';
import { CEPErrorCode } from './model/cep-error-code';
import { CEPError } from './model/cep-error';
import { BASE_URL } from './model/constantes';
import { validarCEP, validarEndereco } from './utils';
import { map, switchMap } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
export class NgxViacepService {
    constructor(http) {
        this.http = http;
    }
    /**
     * Busca o endereço a partir do CEP
     *
     * @param cep
     */
    buscarPorCep(cep) {
        return of(cep).pipe(validarCEP(), switchMap((cepValido) => this.http.get(`${BASE_URL}/${cepValido}/json`)), map((endereco) => {
            if ('cep' in endereco) {
                return endereco;
            }
            throw new CEPError(CEPErrorCode.CEP_NAO_ENCONTRADO);
        }));
    }
    /**
     * Faz a busca aproximada
     *
     * @param uf
     * @param municipio
     * @param logradouro
     */
    buscarPorEndereco(uf, municipio, logradouro) {
        return of({ uf, municipio, logradouro }).pipe(validarEndereco(), switchMap(() => this.http.get(`${BASE_URL}/${uf}/${municipio}/${logradouro}/json`)));
    }
}
NgxViacepService.ɵprov = i0.ɵɵdefineInjectable({ factory: function NgxViacepService_Factory() { return new NgxViacepService(i0.ɵɵinject(i1.HttpClient)); }, token: NgxViacepService, providedIn: "root" });
NgxViacepService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
NgxViacepService.ctorParameters = () => [
    { type: HttpClient }
];
